### ISSUE TITLE
KLB import

### ISSUE THREAD

**Issue by tboo** [bigdataviewer/bigdataviewer-core#20]
It's currently impossible to open KLB stacks with the current version of bigdataviewer. There was a major KLB update recently. I am not sure if it has something to do with it. 

bigdataviewer_fiji-2.0.0.jar
bigdataviewer-core-2.1.0.jar
klb-0.3.4.jar
klb-bdv-0.2.0-SNAPSHOT.jar

**Comment by bhoeckendorf** [comment]
If this used to work for you and you can in principle read klb files with ImageJ/Fiji, then I suspect your intuition that the latest update must have broken something is correct. Although it is a little bit strange that it fails silently with no error message.

Could I ask you to recreate your bigdataviewer xml with the latest version of the klb support and check whether you can browse your images? If so, we know for sure that it is a backwards compatibility issue.

**Comment by tboo** [comment]
I did recreate the XML and the issue remains the same. I get a blank image. The time points are there but the navigation coordinate system also does not appear. 
bigdataviewer works just fine with HDF5 data sets.

**Comment by tpietzsch** [comment]
@tboo can you share a small dataset for testing?

**Comment by tboo** [comment]
Sure. Here's the [link](https://idisk.mpi-cbg.de/~albrecht/TM000001.zip) to a single time point, 2-channel stack.

Thanks a lot for looking into this.

**Comment by bhoeckendorf** [comment]
I think we have 2 unrelated issues here.

1.
You can not open your old xmls anymore because of a backwards compatibility issue in the klb backend. We haven't stumpled upon such an issue here, so I'm not entirely sure what is going on. Your test data does not contain a bigdataviewer xml, could I ask you to provide one?

2.
Before the last update, the klb backend would pad your time series with "missing views" from time point 0 to the first time point with actual data. Since the update, it does not do that anymore. It seems that bigdataviewer can only initialize properly if time point id 0 is configured, either with actual data or as a missing view. This is the reason why you can not open the xml that was generated after the klb backend update.

@tpietzsch - shall I simply reenable the missing views padding or are you going to consider this an issue in bigdataviewer that should be changed? Either way is fine with me.

**Comment by tpietzsch** [comment]
@bhoeckendorf I would consider this an issue with bdv-core. I'll change it to just use the first timepoint for which there is data.

**Comment by tboo** [comment]
Here is the BDV XML.

[BDV.xml.zip](https://github.com/bigdataviewer/bigdataviewer-core/files/293732/BDV.xml.zip)

**Comment by bhoeckendorf** [comment]
@tboo I just noticed the xml you provided is the one that was generated after the update. Do you have an example from before the update?

**Comment by tboo** [comment]
So I just tested the data set with klb-0.3.1 - which as expected worked. Speaking of I can open the data set in BDV. However, the XML I create with BDV is unusable for some reason as it throws an exception when I try to open the XML with BDV again - but that seems to be a totally different issue.
I attached the xml created with klb-0.3.1 (as a txt) regardless

[BDV_old_KLB.txt](https://github.com/bigdataviewer/bigdataviewer-core/files/295129/BDV_old_KLB.txt)
 regardless.

**Comment by bhoeckendorf** [comment]
@tboo - The issue is not with klb (basic Java native interface to klb files), but with klb-bdv (the klb-based BigDataViewer backend). To roll back and recreate the situation before the update, you'd have to downgrade them both. Hence the issue you ran into.

I was wondering whether you could post the actual xml that originally caused you to file this bug. If you can read the individual klb files this xml accesses, and their location in the file system has not changed, then we can safely assume this is not the cause for the issue. So I won't need any image data, but the xml would be quite helpful.

**Comment by bhoeckendorf** [comment]
@tpietzsch - Thanks for looking into the first time point id issue.

**Comment by bhoeckendorf** [comment]
@tboo - I think I have found and fixed a bug in the KLB backend that you ran into. I've released an update that should solve the problem. Please let me know if not. Thanks for letting me know about this and sorry you couldn't browse your data for a few days. I should mention that I've been pretty sick for most of this week, and hence haven't been particularly efficient at debugging.

**Comment by bhoeckendorf** [comment]
@tpietzsch - The problem with a non-configured time point id 0 may be nonexistent after all. I remember starting to prepend missing views, if needed, to avoid problems with that, but these problems are definetely gone. Instead, I had since introduced a different bug in the KLB backend.

However, I noticed that if your actual time points are 5-9, BigDataViewer will display them as 0-4. I'm wondering whether you agree that it may be nicer to consistently use the time point ids of the xml. For the KLB backend, in which datasets are by default spread across multiple files, the implication is that the displayed id, its corresponding entry in the xml, and the file name index are identical, which would be convenient. That being said, if there is a reason to display 0-based time point ids, I'm fine with that.

**Comment by tboo** [comment]
@bhoeckendorf - I can confirm that the KLB import in BDV is working nicely again. Thanks a lot for fixing this. And no worries about the 'delay' - It was a pretty fast fix in my mind. 

@tpietzsch - I can confirm that BDV always starts at t=0 and shifts the actual imported time points accordingly as @bhoeckendorf observed.

**Comment by tpietzsch** [comment]
@bhoeckendorf You are right, it wouId be better to display timepoint ids of the XML. I'm not sure though if these are always available. Using BDV programmatically (as a component library to display stuff) one can just display arbitrary `Source`s that most likely don't come from an XML. But sure would be nice to display ids of the XML if they are available. I'll look into it

**Comment by tpietzsch** [comment]
@bhoeckendorf This is changed now in master. Timepoint ids of the XML are displayed. I'll report here, when a new version has been released...
